### PR TITLE
Attestation data string base types definition (WIP)

### PR DESCRIPTION
**Pull request by djvs**

### CODE REVIEW

File: src/AttestationData.ts [hunk 1]
```diff
@@ -0,0 +1,408 @@
...
+    end_date?: TDate
+
+    // Different levels of generality - zero-indexed, with increasing numbers indicating less generality, to allow for unlimited levels of depth (in practice, 3-5 should be sufficient for most cases).  This allows for arbitrary levels or amounts of generality within sub-attestations, to promote an attestation subject's ability to partially disclose the amount of data provided in an attestation.
+    generality?: number
+
+    // ...extensible with other fields that summarize the content of the attestation - e.g., a list of addresses, accounts, totals of statistics, etc.
+  }
+
+  // Core attestation data, dependent on attestation type
+  data?: Array<Object | string | number> | Object | string | number
```
Comment (edhedges) [inline src/AttestationData.ts]: If we know the supported types for data as seen below and they all look like `TSomeType | Array<TSomeType>`, could we make `IBaseAtt` generic via `IBaseAtt<TAttestation>` and then have data be of type `data?: Array<TAttestation> | TAttestation` where `TAttestation` is `TSomeType`?

File: src/AttestationData.ts [hunk 2]
```diff
@@ -0,0 +1,408 @@
+type TContextField = string | {type: string; data: string}
+
+///////////////////////////////////////////////////
+// Base attestation dataStr type
+///////////////////////////////////////////////////
+export interface IBaseAtt {
+  '@context': TContextField | Array<TContextField>
+  // Date/time attestation was performed
+  date?: string
```
Comment (JaceHensley) [inline src/AttestationData.ts]: Should this be `TDate`? Or maybe a comment for what format the date will be in

File: src/AttestationData.ts [hunk 3]
```diff
@@ -0,0 +1,408 @@
...
+
+  // Secondary provider
+  provider?: string
+
+  // Unique ID for attestation subject from provider
+  id?: number | string
+
+  // Date/time when attestation should be considered void (e.g., for credential expiry)
+  expiry_date?: string
```
Comment (JaceHensley) [inline src/AttestationData.ts]: See above for `date`

File: src/AttestationData.ts [hunk 4]
```diff
@@ -0,0 +1,408 @@
...
+
+  // Optional summary object (often useful for multiple-account types - otherwise, the data field is preferable for most of these fields, for simplicity's sake
+  summary?: {
+    // Single date/time during which attestation was applicable
+    date?: TDate
```
Comment (JaceHensley) [inline src/AttestationData.ts]: Will it ever be `date` and `start_date`+`end_date` or just `date` OR `start_date`+`end_date`.

If it's the latter maybe this type should be:

```
type TBaseSummary = {
  generality?: number
}

type TSummaryWithDate = TBaseSummary & {
  date: TDate
}

type TSummaryWithStartEndDate = TBaseSummary & {
  start_date: TDate
  end_date: TDate
}

type TSummary = TSummaryWithDate | TSummaryWithStartEndDate
```

File: src/AttestationData.ts [hunk 5]
```diff
@@ -0,0 +1,408 @@
...
+export interface IBaseAttPhone extends IBaseAtt {
+  data: TBaseAttPhoneData | Array<TBaseAttPhoneData>
+}
+
+///////////////////////////////////////////////////
+// Email attestation dataStr type
+///////////////////////////////////////////////////
+export interface IBaseAttEmailData {
+  email?: string
+  start_date?: string
```
Comment (JaceHensley) [inline src/AttestationData.ts]: Why does this one have `start/end_date`

File: src/AttestationData.ts [hunk 6]
```diff
@@ -0,0 +1,408 @@
...
+
+///////////////////////////////////////////////////
+// ID document attestation dataStr type
+///////////////////////////////////////////////////
+export interface IBaseAttIDDocumentData {
+  date: TDate
+  name: TPersonalName
+  country: string
+
+  // Primarily modelled after KYC2020 responses, most fields left optional for flexibility
```
Comment (JaceHensley) [inline src/AttestationData.ts]: All the fields in this interface seem to be required

File: src/AttestationData.ts [hunk 7]
```diff
@@ -0,0 +1,408 @@
...
+    // Start date/time of period during which attestation was applicable
+    start_date?: TDate
+
+    // End date/time of period during which attestation was applicable
+    end_date?: TDate
+
+    // Different levels of generality - zero-indexed, with increasing numbers indicating less generality, to allow for unlimited levels of depth (in practice, 3-5 should be sufficient for most cases).  This allows for arbitrary levels or amounts of generality within sub-attestations, to promote an attestation subject's ability to partially disclose the amount of data provided in an attestation.
+    generality?: number
+
+    // ...extensible with other fields that summarize the content of the attestation - e.g., a list of addresses, accounts, totals of statistics, etc.
```
[the remark this answers is not in the document]
Comment (djvs) [reply src/AttestationData.ts]: It's arbitrarily extensible - this entire specification is essentially a suggestion to increase the likelihood that an attestation will adhere to a given expectation, which is why most fields are optional.  Actual implementations are expected to extend the base specifications.

File: src/AttestationData.ts [hunk 8]
```diff
@@ -0,0 +1,408 @@
...
+// Name attestation dataStr type
+///////////////////////////////////////////////////
+export interface IBaseAttName extends IBaseAtt {
+  data: TPersonalName | Array<TPersonalName>
+}
+
+///////////////////////////////////////////////////
+// SSN/government ID # attestation dataStr type
+///////////////////////////////////////////////////
+export interface IBaseAttSSNData {
```
[the remark this answers is not in the document]
Comment (djvs) [reply src/AttestationData.ts]: Yes, considering deprecating this in favor of IBaseAttAccount.

File: src/AttestationData.ts [hunk 9]
```diff
@@ -0,0 +1,408 @@
...
+  postal_code: string | number
+  region_1: string
+  region_2?: string
+  country?: string
+}
+
+///////////////////////////////////////////////////
+// Phone attestation dataStr type
+///////////////////////////////////////////////////
+export type TBaseAttPhoneData = string | TPhoneNumber
```
[the remark this answers is not in the document]
Comment (djvs) [reply src/AttestationData.ts]: Deleted

File: src/AttestationData.ts (hunk 1 above)
Comment (djvs) [reply src/AttestationData.ts]: There will be more than one exception (currently I think just DOB).  I'm considering using string types inside the top-level attestation specification to force an attestation type's adherence to a given structure.

File: src/AttestationData.ts (hunk 2 above)
Comment (djvs) [reply src/AttestationData.ts]: Yes, TDate was introduced halfway through the PR, fixed

File: src/AttestationData.ts (hunk 6 above)
Comment (djvs) [reply src/AttestationData.ts]: Not following

File: src/AttestationData.ts (hunk 5 above)
Comment (djvs) [reply src/AttestationData.ts]: Sometimes emails associated with an individual or organization have a beginning or expiry date.  Just so happens I had that exact problem this morning.  E.g., a company issuing an attestation for when an employee has access to an email.

File: src/AttestationData.ts (hunk 4 above)
Comment (djvs) [reply src/AttestationData.ts]: That's at the base type for all attestations, so it could certainly be one of start_date or end_date without the other.

**Pull request merged**